internal/controllers/gateway/httproute_controller.go

```diff
@@ -585,3 +600,101 @@ func (r *HTTPRouteReconciler) getHTTPRouteRuleReason(ctx context.Context, httpRo
+func (r *HTTPRouteReconciler) ensureParentsAcceptedCondition(
+	ctx context.Context,
+	httproute *gatewayv1beta1.HTTPRoute,
+	gateways []supportedGatewayWithCondition,
+	conditionStatus metav1.ConditionStatus,
+	conditionReason string,
+	conditionMessage string,
+) (bool, error) {
+	// map the existing parentStatues to avoid duplications
+	parentStatuses := make(map[string]*gatewayv1beta1.RouteParentStatus)
+	for _, existingParent := range httproute.Status.Parents {
+		namespace := httproute.Namespace
+		if existingParent.ParentRef.Namespace != nil {
+			namespace = string(*existingParent.ParentRef.Namespace)
+		}
+		existingParentCopy := existingParent
+		var sectionName string
+		if existingParent.ParentRef.SectionName != nil {
+			sectionName = string(*existingParent.ParentRef.SectionName)
+		}
+		parentStatuses[fmt.Sprintf("%s/%s/%s", namespace, existingParent.ParentRef.Name, sectionName)] = &existingParentCopy
+	}
+
+	statusChanged := false
+	for _, g := range gateways {
+		gateway := g.gateway
+		parentRefKey := fmt.Sprintf("%s/%s/%s", gateway.Namespace, gateway.Name, g.listenerName)
+		parentStatus, ok := parentStatuses[parentRefKey]
+		if ok {
+			// update existing parent in status.
+			changed := updateAcceptedConditionInRouteParentStatus(parentStatus, conditionStatus, conditionReason, conditionMessage, httproute.Generation)
+			statusChanged = statusChanged || changed
+		} else {
+			// add a new parent if the parent is not found in status.
+			newParentStatus := &gatewayv1beta1.RouteParentStatus{
+				ParentRef: gatewayv1beta1.ParentReference{
+					Namespace:   (*gatewayv1beta1.Namespace)(pointer.String(gateway.Namespace)),
+					Name:        gatewayv1beta1.ObjectName(gateway.Name),
+					SectionName: (*gatewayv1beta1.SectionName)(pointer.String(g.listenerName)),
+					// TODO: set port.
+				},
+				Conditions: []metav1.Condition{
+					{
+						Type:               string(gatewayv1beta1.RouteConditionAccepted),
+						Status:             conditionStatus,
+						ObservedGeneration: httproute.Generation,
+						LastTransitionTime: metav1.Now(),
+						Reason:             conditionReason,
+						Message:            conditionMessage,
+					},
+				},
+			}
+			httproute.Status.Parents = append(httproute.Status.Parents, *newParentStatus)
+			parentStatuses[parentRefKey] = newParentStatus
+			statusChanged = true
+		}
+	}
+
+	// update status if needed.
+	if statusChanged {
+		if err := r.Status().Update(ctx, httproute); err != nil {
+			return false, err
+		}
+		return true, nil
+	}
+	// no need to update if no status is changed.
+	return false, nil
+}
+
+// updateAcceptedConditionInRouteParentStatus updates conditions with type "Accepted" in parentStatus.
+// returns true if the parentStatus modified.
+func updateAcceptedConditionInRouteParentStatus(
+	parentStatus *gatewayv1beta1.RouteParentStatus,
+	conditionStatus metav1.ConditionStatus,
+	conditionReason string,
```

Comment:
Can we use `gatewayv1beta1.RouteConditionReason` for this argument? It will avoid us make a type cast in e.g. `internal/controllers/gateway/httproute_controller.go:331` and make this more type safe ( no accidental usage of message in place of reason because they are the same type ). WDYT?

```suggestion
	conditionReason gatewayv1beta1.RouteConditionReason,
```

Reply:
done